File: Level 2/삼각 달팽이/solution.py
from itertools import chain


def solution(n):
    pyramid = [[0] * (i + 1) for i in range(n)]
    num = sum(len(row) for row in pyramid)
    direction = "down"
    x, y = 0, 0

    for i in range(1, num + 1):
        pyramid[y][x] = i

        if direction == "down":
            if y + 1 < n and pyramid[y + 1][x] == 0:
                y += 1
            else:
                x += 1
                direction = "right"
        elif direction == "right":
            if x + 1 < n and pyramid[y][x + 1] == 0:
                x += 1
            else:
                x -= 1
                y -= 1
                direction = "up"
        elif direction == "up":
            if x - 1 >= 0 and y - 1 >= 0 and pyramid[y - 1][x - 1] == 0:
                x -= 1
                y -= 1
            else:
                y += 1
                direction = "down"

    return list(chain(*pyramid))

File: Level 2/삼각 달팽이/solution.js
function solution(n) {
  let num = 0;

  const pyramid = Array.from({ length: n }, (_, index) => {
    const rowLength = index + 1;
    num += rowLength;
    return new Array(rowLength).fill(0);
  });

  let direction = "down";
  let [x, y] = [0, 0];

  for (let i = 1; i <= num; i++) {
    pyramid[y][x] = i;

    switch (direction) {
      case "down":
        if (y + 1 < n && pyramid[y + 1][x] === 0) {
          y++;
        } else {
          x++;
          direction = "right";
        }
        break;
      case "right":
        if (x + 1 < n && pyramid[y][x + 1] === 0) {
          x++;
        } else {
          x--;
          y--;
          direction = "upLeft";
        }
        break;
      case "upLeft":
        if (x - 1 >= 0 && y - 1 >= 0 && pyramid[y - 1][x - 1] === 0) {
          x--;
          y--;
        } else {
          y++;
          direction = "down";
        }
        break;
    }
  }

  return pyramid.flat();
}
